src/http.nim

```diff
@@ -0,0 +1,250 @@
+
+type
+  HttpClient* = ref object
+    connected: bool
+    currentURL: Uri      ## Where we are currently connected.
+    headers: HttpHeaders ## Headers to send in requests.
+    transp*: StreamTransport
+    buf: seq[byte]
+
+  HttpHeaders* = object
+    table*: TableRef[string, seq[string]]
+
+  ReqStatus = enum
+    Success, Error, ErrorFailure
+
+  AsyncCallback = proc (transp: StreamTransport,
+      header: HttpRequestHeader): Future[void] {.closure, gcsafe.}
+  HttpServer* = ref object of StreamServer
+    callback: AsyncCallback
+
+proc recvData(transp: StreamTransport): Future[seq[byte]] {.async.} =
+  var buffer = newSeq[byte](MaxHttpHeadersSize)
+  var error = false
+  try:
+    let hlenfut = transp.readUntil(addr buffer[0], MaxHttpHeadersSize,
+        sep = HeaderSep)
+    let ores = await withTimeout(hlenfut, HttpHeadersTimeout)
+    if not ores:
+      # Timeout
+      debug "Timeout expired while receiving headers",
+             address = transp.remoteAddress()
+      error = true
+    else:
+      let hlen = hlenfut.read()
+      buffer.setLen(hlen)
+  except TransportLimitError:
+    # size of headers exceeds `MaxHttpHeadersSize`
+    debug "Maximum size of headers limit reached",
+          address = transp.remoteAddress()
+    error = true
+  except TransportIncompleteError:
+    # remote peer disconnected
+    debug "Remote peer disconnected", address = transp.remoteAddress()
+    error = true
+  except TransportOsError as exc:
+    debug "Problems with networking", address = transp.remoteAddress(),
+          error = exc.msg
+    error = true
+
+  if error:
+    buffer.setLen(0)
+  return buffer
+
+proc newConnection(client: HttpClient, url: Uri) {.async.} =
+  if client.connected:
+    return
+
+  let port =
+    if url.port == "": 80
+    else: url.port.parseInt
+
+  client.transp = await connect(initTAddress(url.hostname, port))
+
+  # May be connected through proxy but remember actual URL being accessed
+  client.currentURL = url
+  client.connected = true
+
+proc generateHeaders(requestUrl: Uri, httpMethod: string,
+    additionalHeaders: HttpHeaders): string =
+  # GET
+  var headers = httpMethod.toUpperAscii()
+  headers.add ' '
+
+  if not requestUrl.path.startsWith("/"): headers.add '/'
+  headers.add(requestUrl.path)
+
+  # HTTP/1.1\c\l
+  headers.add(" HTTP/1.1" & CRLF)
+
+  for key, val in additionalHeaders.table:
+    headers.add(key & ": " & val.join(", ") & CRLF)
+  headers.add(CRLF)
+  return headers
+
+# Send request to the client. Currently only supports HTTP get method.
+proc request*(client: HttpClient, url, httpMethod: string,
+             body = "", headers: HttpHeaders): Future[seq[byte]]
+             {.async.} =
+  # Helper that actually makes the request. Does not handle redirects.
+  let requestUrl = parseUri(url)
+  if requestUrl.scheme == "":
+    raise newException(ValueError, "No uri scheme supplied.")
+
+  await newConnection(client, requestUrl)
```

Comment:
```suggestion
  await client.connect(requestUrl)
```

For reference, Nim supports UFFC - https://en.wikipedia.org/wiki/Uniform_Function_Call_Syntax